onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,719 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/propagate_cast_ops.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/utils.h"
+#include <deque>
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::common;
+namespace onnxruntime {
+// fp16_allow_ops do not computationally modify the float/float16 input to produce output, such as transpose.
+// fp16_safe_ops have cuda kernal implementation that first converts the input to float and result back to float.
+// The opcodes are split into multiple levels. Cast propagation is done based on the level. Level 2 op code
+// list includes Level 1 list also.
+static std::vector<std::unordered_set<std::string>> fp16_allow_ops = {
```

Comment:
Personally, I found this complicates our optimization levels... 
My original assumption is that, fp16_allow_ops  == Level1 and fp16_safe_ops  == Level2 . My recommendation is that we can just follow the level optimization in apex.amp. https://nvidia.github.io/apex/amp.html 

Level 0: don't do cast propagation
Level 1: Recommended for typical use, this only contains the ops that **DOESN'T change the numerical results.** 
- Ops involve no-computation, such as "Transpose, Concat, Split". 
- Ops involve simple and fp16-safe computation, e.g. "Relu"
- Ops that has "Input-CastTo32, Compute in fp32, CastTo16-Output" kind of CUDA kernel, e.g. Gelu, FastGelu. (They are equivalent a Fused op of "Cast-Gelu-Caste"

Level2: If you need more perf, but "numerical results" could be affected. 
- Ops that change the computation precision in its kernel, e.g. LayerNorm(input output are not directly cast to fp32, but mean/variance are stored in fp32)

Reply:
There can be a dynamic boarder between Level 1 and Level 2, as we experiment on more models. 
If we found any op always safe, we can promote it to Level 1. 
If we found any op not safe, we can demote it to level 2.